### DISCUSSION TITLE
Is there any option to calculate Berry connections ?

### DISCUSSION THREAD

**Discussion by AbhirupRK** [wannier-berri/wannier-berri#223] [Q&A] [answered]
I am wondering if there is any way to get $\< m | \partial_k | n \>$ elements. Can I get it from .amn file ?

**Comment by stepan-tsirkin** [comment, marked as answer]
the .amn has nothing to do with Berry connections, is contains the projections of the ab initio Bloch states to the trial orbitals. 

you may look at this PR :  [ tabulation of Berry connection #184 ](https://github.com/wannier-berri/wannier-berri/pull/184) 

Also, you may access them as `A_H` property og the `Data_K` object https://github.com/wannier-berri/wannier-berri/blob/27a451f7e1e7890f9256ba0a142cf5de1c18003f/wannierberri/data_K.py#L436-L439

**Reply by AbhirupRK** [reply]
How can I use this PR ? Do I need to merge this with the latest version ?

**Reply by stepan-tsirkin** [reply]
you may just install from the corresponding branch  

I have just updated it with the latest main branch

**Reply by AbhirupRK** [reply]
Thanks a lot @stepan-tsirkin. I will try it out.

**Reply by AbhirupRK** [reply]
@stepan-tsirkin, that worked perfectly. Can I get derivative ($\partial_k$) of this quantity ? Or is there a way to get derivative of any implemented quantity ?

**Reply by stepan-tsirkin** [reply]
@AbhirupRK , not so far, but the derivatives of Berry connection will be there soon. I will let you know.

**Reply by AbhirupRK** [reply]
Thanks a lot.

**Comment by AbhirupRK** [comment]
Dear @stepan-tsirkin ,

Can you please show me an example on how to use the `Data_K` object ? Specifically, I need the Berry connection property.

Previousy, I used the PR you have mentioned. However, I'm getting lower value (factor of Pi) after performing an integral.

**Reply by stepan-tsirkin** [reply]
Well, it is somethin like this:
``` python
system = ...
grid = ...
datak=  wannierberri.data_K.Data_K(system, dK=[0,0,0], grid=grid,  **parameters)
berry_connection = datak.A_H
```
Or could you specify the question.

Also, what integral are you performing and what are you comparing?

**Reply by AbhirupRK** [reply]
Thanks @stepan-tsirkin, you just gave what I asked for. However, I don't think it will solve my problem.

Since you asked, I'm stating my problem specifically :

I am trying to reproduce `Fig. 6(e)` of the paper [npj Quantum Materials (2020) 5:96](https://doi.org/10.1038/s41535-020-00298-y) using `Eq. (5)`.

My result came out to be exactly the same except for a **factor of 10**. i.e., upon multiplying my result with 10 correctly reproduces the fugure of the paper.

This multiplication factor could be 10, $\pi^2$ or even $4\pi$ (as there is always some numerical accuracy issue, can't tell exactly).

The factor issue is not with the outside coefficients (checked in other ways). It's coming from the integral. I used `Berry connection` and `Velocity` from `wannierberri`.

**Reply by stepan-tsirkin** [reply]
The assumed way for evaluating such integral is by creating a Calculator, somethinf by analogy with optical conductivity https://github.com/wannier-berri/wannier-berri/blob/dd6aeaf39311aa3747e520eb5378aea064b8b007/wannierberri/calculators/dynamic.py#L166-L197

the internal units of wannierberri are eV and  angstroms, so berry connection is in $\mathring{A}^{-1}$, velocity in eV $\cdot\mathring{A}$

But i do not know where you (or they) lost a factor.

**Reply by AbhirupRK** [reply]
Well, I simply tabulated all the required data and obtained the integral using `simpson` method of `scipy`. Need to figure out the mistake.

**Reply by AbhirupRK** [reply]
> the internal units of wannierberri are eV and angstroms, so berry connection is in $\mathring{A}^{-1}$, velocity in eV $\cdot\mathring{A}$

Berry connection is in $\mathring{A}$ , right ?

**Reply by stepan-tsirkin** [reply]
Right, in is Angstrom, not inverse.

**Reply by AbhirupRK** [reply]
Hi, can I get $\nabla E$ from `DataK` ?

I used `datak.delE_K`, but it is not same as `wberri.calculators.tabulate.Velocity()`.

**Reply by stepan-tsirkin** [reply]
I think it should be the same, at least for non-degenerate states. Maybe there is a different treatment of degeneracies, but for degenerate states velocities of individual bands are not  reallly defined.  Are they much different?

**Reply by AbhirupRK** [reply]
Well as I checked it so far, they are different at Gamma point (and sometimes at other kpoints). But for rest of the kpoints, they are exactly the same.

**Reply by stepan-tsirkin** [reply]
That sounds like those are special points with degeneracies, so maybe a different threshold to consider bands as degenerate?

**Reply by AbhirupRK** [reply]
Well, for now I'm using the `Velocity()` only.
